dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -246,136 +172,292 @@ bool tryResetPeerIdInRegion(RegionPtr region, const RegionLocalState & region_st
-                    {
-                        success = true;
-                        LOG_INFO(
-                            log,
-                            "Select checkpoint with seq {} from store {} takes {} seconds, candidate_store_id size {} "
-                            "[region_id={}]",
-                            data_seq,
-                            checkpoint_info->remote_store_id,
-                            watch.elapsedSeconds(),
-                            candidate_store_ids.size(),
-                            region_id);
-                        GET_METRIC(tiflash_fap_task_duration_seconds, type_build_stage).Observe(watch.elapsedSeconds());
-                        break;
-                    }
-                    else
-                    {
-                        LOG_DEBUG(
-                            log,
-                            "Checkpoint with seq {} from store {} doesn't contain reusable region info [region_id={}]",
-                            data_seq,
-                            store_id,
-                            region_id);
-                    }
+                        store_id);
                 }
             }
-            if (!success)
+        }
+        {
+            if (watch.elapsedSeconds() >= settings.fap_wait_checkpoint_timeout_seconds)
             {
-                if (watch.elapsedSeconds() >= settings.fap_wait_checkpoint_timeout_seconds)
-                {
-                    GET_METRIC(tiflash_fap_task_result, type_failed_timeout).Increment();
-                    return genFastAddPeerRes(FastAddPeerStatus::NoSuitable, "", "");
-                }
-                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+                // TODO(fap) Cancel and remove from AsyncTasks
+                // This could happen if there are too many pending tasks in queue,
+                LOG_INFO(log, "FastAddPeer timeout region_id={} new_peer_id={}", region_id, new_peer_id);
+                GET_METRIC(tiflash_fap_task_result, type_failed_timeout).Increment();
+                return genFastAddPeerRes(FastAddPeerStatus::NoSuitable, "", "");
             }
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
         }
-        after_build();
+    }
+}
+
+FastAddPeerRes FastAddPeerImplWrite(
+    TMTContext & tmt,
+    UInt64 region_id,
+    UInt64 new_peer_id,
+    CheckpointRegionInfoAndData && checkpoint,
+    UInt64 start_time)
+{
+    auto * log = &Poco::Logger::get("FastAddPeer");
+    auto fap_ctx = tmt.getContext().getSharedContextDisagg()->fap_context;
+    const auto & settings = tmt.getContext().getSettingsRef();
+
+    Stopwatch watch;
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_duration_seconds, type_write_stage).Observe(watch.elapsedSeconds()); });
+    GET_METRIC(tiflash_fap_task_state, type_writing_stage).Increment();
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_writing_stage).Decrement(); });
```

Comment:
The "write" means it will actually write some information related to the region/segment which will be added(ingested) into the formal data. So we have to clean it first.